### PR TITLE
chore(iam): add a `PrecreatedRole` class

### LINKED ISSUE #22750: Generate report of IAM policies that would have been created
https://github.com/aws/aws-cdk-rfcs/blob/master/text/0063-precreated-roles.md

- When operating in "no role creation" mode, roles will synthesize their policy documents to a report file.
- We should be able to detect new iam.Policy() as well, as I believe it calls role.attachPolicy(). We record the policy and prevent its synthesis of CfnPolicy.
  - I'm not sure we will be able to detect new iam.ManagedPolicy().

### PR DESCRIPTION
**Pull request by corymhall**
This PR adds a new private class for creating a `PrecreatedRole`. For background see the [precreated-roles
RFC](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0063-precreated-roles.md).

This PR adds the class and the functionality to generate a report as part of synthesis, but it does not yet expose this publicly. A future PR will use this class as part of the `Role` class.

The report that is generated will contain references to resources that have not yet been created. Those values are not easy to read in a report and make it harder to search/replace with a hard coded value once the resource _is_ created. To account for this we replace any references with easier to read values. For example, this:

```
"Resource": {
  "Fn::Join": [
    "",
    [
      "arn:",
      {
        "Ref": "AWS::Partition"
      },
      ":iam::",
      {
        "Ref": "AWS::AccountId"
      },
      ":role/Role"
    ]
  ]
}
```

Will become:

`"Resource": "arn:(PARTITION):iam::(ACCOUNT):role/Role"`

note: labeled as `chore` since this does not add any public facing API. Future PRs will add the public API that will expose the report generation

closes #22750

----

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (Naumel) [inline packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:82]: [nitpick] The goal here is to fill in the value. If the Role is created and the name is `missing role` it makes it a bit extra effort to troubleshoot/ audit.
Should it be a `Default`+timestamp?

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (Naumel) [inline packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:162]: Note: wrt the debate of the LogicalID that's generated via CFn, as a recommended best practice vs the 2-step deploy, grab the ARNs, fill them in the policy.

Scenario: deploy the roles, grab the ARNs for the resources, fill in the policies with the actual values, add them to the prod stack to the role for explicit and complete boundary.

Q1: A deployment that changes the Logical ids, will it also invalidate the policies? Would they have to be re-edited?
Q2: Can the teams front-load some of the work by having a list of resources with fixed logicalIds and know that they will have to 2-step deploy if names change? Does the list of resources make it more transparent which policy will be rendered invalid too (if the assumption is correct)?

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (Naumel) [inline packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:218]: [nit] I think we could have more than one role with `missing role` name in the report.
Do you think the context offers enough information for the customer's report readability?

**Review by Naumel — APPROVED** [review]
Looks good, with a couple of scenarios and questions I'd like to get clarity on (therefore added tag).

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (rix0rrr) [reply packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:162]: * Q1: yes. If logical IDs change, resources must be replaced which requires a name change (hence invalidating the policies)
* Q2: Not sure I understand the Q.

I think the nature of this method brings with it that you're going to have to wildcard your resources. We're not changing that. The only thing we're doing is adding more information that you otherwise would have had.

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:164]: Shouldn't this be the policysynthesizer's job? Feels like part of rendering.

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:257]: Why is this important?

If the goal is to not update the report unless synthesis completes successfully, I think that's more appropriately a part of the synthesizer protocol (to be able to explicitly request this behavior) than trying to employ your knowledge of the CDK lifecycle to do these order-dependent things.

In general, order-dependent coding is scary and dangerous (to me).

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (corymhall) [reply packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:22]: I'll add some more docs, but the idea is the same as what we do for `ImmutableRole` https://github.com/aws/aws-cdk/blob/547ba5f1159f1e8a91072dfb0484c6b294994b2c/packages/@aws-cdk/aws-iam/lib/private/immutable-role.ts#L32

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (corymhall) [reply packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:164]: yeah i'll move it. For some reason I forgot I can use `Tokenization.resolve` instead of `Stack.resolve`  ha

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (corymhall) [reply packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts]: yeah I don't think it's actually important that this be _attached_ at a specific point since really all we care about is that the synthesizer runs as part of synthesis, which it will regardless of when we attach it. I've removed the `addValidation` piece.

File: packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts
Comment (Naumel) [reply packages/@aws-cdk/aws-iam/lib/private/precreated-role.ts:257]: Q2 was an after thought for an edge case, let me add context:

If there is a customer with a security team enforcing the condition not to wildcard anything, then both roles and policies' content could be done with one change if the logical ids (names) are known beforehand.

Probably the question would be: is there a case that would justify hardcoding names in order to unblock work? 

Thinking about it, the answer becomes more and more `no`. As any issue with such a setup would get a swift reference to the page containing best practices and recommendations :-)

**Pull request merged**
